pkg/resource/deploy/step_generator.go

```diff
@@ -615,7 +611,25 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, res
 		return []Step{NewImportStep(sg.deployment, event, new, goal.IgnoreChanges, randomSeed)}, nil
 	}
 
-	isTargeted := sg.isTargetedForUpdate(new)
+	isUserResource := !(providers.IsDefaultProvider(urn) || urn.Type() == resource.RootStackType)
+
+	// Internally managed resources are under Pulumi's control and changes or creations should be invisible
+	// to the user.
+
+	// Resources are targeted by default
+	isTargeted := true
+	if sg.isTargetedUpdate() && isUserResource {
+		if !sg.replaceTargetsOpt.IsConstrained() {
+			// Not a replace, so check if the resource is targeted for an update.
+			isTargeted = sg.isTargetedForUpdate(new)
+		} else if !sg.updateTargetsOpt.IsConstrained() {
+			// Not an update, so check if the resource is targeted for a replace.
+			isTargeted = sg.isTargetedReplace(urn)
+		} else {
+			return nil, result.Error(
+				"update is marked constrained, but is not constrained to an update or a replace")
+		}
+	}
```

Comment:
This logic is very miserable because if we add another target option i.e. `--target-foo` the checks will explode combinatorially.

Reply:
Why isn't this just `sg.isTargetedForUpdate(new) || sg.isTargetedReplace(urn)`?

Reply:
ick, ok something to think about